shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs

```diff
@@ -1103,3 +1106,18 @@ updateNES
   bcur
   ls =
     NewEpochState eL bprev bcur (EpochState acnt ss ls pr pp nm) ru pd
+
+returnRedeemAddrsToReserves ::
+  ShelleyBased era =>
+  EpochState era ->
+  EpochState era
+returnRedeemAddrsToReserves es = es {esAccountState = acnt', esLState = ls'}
+  where
+    ls = esLState es
+    us = _utxoState ls
+    UTxO utxo = _utxo us
+    (redeemers, nonredeemers) = Map.partition (\(TxOut a _) -> isBootstrapRedeemer a) utxo
```

Comment:
Note that pattern matching on a `TxOut` causes it to be decompacted, which is rather expensive when done for the entire UTxO. It would be great if this check could be done without decompacting, it should significantly speed up the Shelley-to-Allegra translation of the ledger.

Reply:
that makes sense, thank you! I will see about doing something smarter.

Reply:
I just talked with @redxaxder , and we think that the smart thing to do is to finish CAD-1874. This ticket will make it so that using the pattern as I did will not do any unnecessary decompactification (is that a word? :).

Reply:
@mrBliss  this is CAD-1874 PR: https://github.com/input-output-hk/cardano-ledger-specs/pull/1935

Reply:
I see. I don't think that PR changes much, as you're still going to decompactify/deserialise the address. You'll avoid decompactifying the value (`Coin`), but I believe that's very cheap compared to decompactifying the address. But I agree, let's merge that PR first before trying to do something clever for `isBootstrapRedeemer` :slightly_smiling_face:.

I'm not familiar with the encoding of addresses, but if it is possible to just check some bits to see whether an address is a boostrap redeemer address, then you wouldn't have to deserialise the whole address (also no allocation).

Reply:
> you're still going to decompactify/deserialise the address.

Could you expand on this?

The purpose of the change is to not decode the parts that are not forced. If it is done successfully, is there still a problem here?

Reply:
Sure: `isBootstrapRedeemer` is implemented as follows:

```haskell
isBootstrapRedeemer :: Addr era -> Bool
isBootstrapRedeemer (AddrBootstrap (BootstrapAddress (Byron.Address _ _ Byron.ATRedeem))) = True
isBootstrapRedeemer _ = False
```

In your #1935:
```haskell
data Addr era
  = Addr Network (PaymentCredential era) (StakeReference era)
  | AddrBootstrap !(BootstrapAddress era)

...

decompactAddr :: forall era. Era era => CompactAddr era -> Addr era
decompactAddr (CompactAddr bytes) = if testBit header byron
  then AddrBootstrap $ {- THIS: -} run 0 bytes getBootstrapAddress 
  else Addr addrNetId paycred stakecred
  ..
```

So the entire `BootstrapAddress` will be decompactified. IIRC, deserialising a `BootstrapAddress` (CBOR in `Binary` or what was it again?) was much more expensive than a regular Shelley address. You could make that lazy too, as we're only interested in the [`addrType` field](https://github.com/input-output-hk/cardano-ledger-specs/blob/7a6131a5fd8009fbced7231f84c7b8e111be05da/byron/ledger/impl/src/Cardano/Chain/Common/Address.hs#L116). An alternative, if the encoding permits, is to implement the `addrType` check as a bit(pattern) check.

Reply:
@mrBliss do you think it is okay to merge this PR now and leave this optimization as a nice to have for later?

Reply:
Yeah, sure! Maybe create an issue somewhere so we don't forget about it. (Or maybe there is already an omnibus ticket about optimisations/performance/...?)